File: src/com/gmail/marcosav2010/communicator/packet/wrapper/PacketReader.java
package com.gmail.marcosav2010.communicator.packet.wrapper;

import java.io.ByteArrayInputStream;

import com.gmail.marcosav2010.communicator.packet.AbstractPacket;
import com.gmail.marcosav2010.communicator.packet.PacketRegistry;
import com.gmail.marcosav2010.communicator.packet.wrapper.exception.PacketReadException;

public class PacketReader {

	public AbstractPacket read(byte[] bytes) throws PacketReadException {
		AbstractPacket packet = null;

		try (ByteArrayInputStream in = new ByteArrayInputStream(bytes); PacketDecoder decoder = new PacketDecoder(in)) {

			byte packetType = decoder.readByte();

			Class<? extends AbstractPacket> packetClass = PacketRegistry.getById(packetType);
			if (packetClass == null)
				throw new PacketReadException("Packet type " + Byte.toUnsignedInt(packetType) + " not recognized");

			packet = packetClass.getConstructor().newInstance();

			packet.decode(decoder);

		} catch (PacketReadException ex) {
			throw ex;
		} catch (Exception ex) {
			throw new PacketReadException(ex);
		}

		return packet;
	}
}

File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/wrapper/PacketReader.kt
package com.gmail.marcosav2010.communicator.packet.wrapper

import com.gmail.marcosav2010.communicator.packet.AbstractPacket
import com.gmail.marcosav2010.communicator.packet.PacketRegistry.getById
import com.gmail.marcosav2010.communicator.packet.wrapper.exception.PacketReadException
import java.io.ByteArrayInputStream

class PacketReader {

    fun read(bytes: ByteArray): AbstractPacket {
        var packet: AbstractPacket

        try {
            ByteArrayInputStream(bytes).use {
                PacketDecoder(it).use { decoder ->
                    val packetType = decoder.readByte()
                    val packetClass = getById(packetType)
                            ?: throw PacketReadException("Packet type ${packetType.toInt()} not recognized")

                    packet = packetClass.getConstructor().newInstance()!!
                    packet.decode(decoder)

                    return packet
                }
            }
        } catch (ex: PacketReadException) {
            throw ex
        } catch (ex: Exception) {
            throw PacketReadException(ex)
        }
    }
}